lib/sru.rb

```diff
@@ -80,4 +80,88 @@ def self.wrap(records)
       </searchRetrieveResponse>
     EOS
   end
+
+  def self.download_all(type)
+    start_doc = fetch(query: config["#{type}_query"])
+    marc_count = Nokogiri::XML(start_doc).css("numberOfRecords").text.to_i
+    Rails.logger.info "Downloading #{marc_count} records (this is slow)"
+
+    batched_count = if marc_count > 10_000
+                      10_000
+                    else
+                      marc_count
+                    end
+
+    batches = batched_count / config[:batch_size]
+    all = Array.new(batches) do |i|
+      start = (i * config[:batch_size]) + 1
+      strip(download_range(type, start, config[:batch_size]))
+    end
+
+    if marc_count > 10_000
+      all << download_extra(type, marc_count)
+    else
+      remainder = marc_count % config[:batch_size]
+      if remainder.positive?
+        all << download_range(type, (marc_count - remainder + 1), remainder)
+      end
+    end
+
+    output = File.join(Settings.marc_directory, "#{type}-metadata.xml")
+    File.open(output, "w") do |f|
+      f.write wrap(all)
+      num = MARC::XMLReader.new(StringIO.new(all)).map { |r| r }.count
+      Rails.logger.info "Wrote #{num} records to #{output}"
+    end
+  end
+
+  # @param [Symbol] type
+  # @param [Integer] start
+  # @param [Integer] number
+  def self.download_range(type, start, number)
+    marc_batch = fetch(query: config["#{type}_query"],
+                       max: number,
+                       start: start)
+
+    output = File.join(
+      Settings.marc_directory,
+      format("%s-%05d-%05d.xml", type, start, (start + number - 1))
+    )
+
+    File.open(output, "w") do |f|
+      f.write marc_batch
+
+      num = MARC::XMLReader.new(StringIO.new(marc_batch)).map { |r| r }.count
+      Rails.logger.info "Wrote #{num} records to #{output}"
+    end
+
+    marc_batch
+  end
+
+  # Work around the fact that when a query matches more than 10,000
+  # records, Alma will only let you access the first 10,000
+  #
+  # @param [Symbol] type
+  # @param [Integer] extra
+  def self.download_extra(type, extra)
+    raise "We've passed 10,000 ETDs!!!" if type == :etd
```

Comment:
I'm not going to bother figuring out the query for individual ETDs until I have to.